src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java

```diff
@@ -0,0 +1,264 @@
+   * @return schema calculated from query
+   * @throws Exception is thrown by httpclientlib
+   */
+  @Path("getSchema")
+  public Schema getSchema(Request request) throws Exception {
+    return SalesforceSchemaUtil.getSchemaFromQuery(new AuthenticatorCredentials(request.username, request.password,
+                                                                                request.clientId, request.clientSecret,
+                                                                                request.loginUrl), request.query);
+  }
+
+  /**
+   * An advanced version of split csv row by comma. We cannot simply split by comma since some values may have comma
+   * inside them and Salesforce does not escape it.
+   *
+   * The format of Salesforce csv row is:
+   * "value1","value2","value3"
+   *
+   * So we are splitting by "\",\"" instead of by comma. Since quotes are not allowed without escaping
+   * in values this works in any case.
+   *
+   *
+   * @param csvRow one row in csv format with quoted values
+   * @return an array of values
+   */
+  private String[] getValuesFromCSVRow(String csvRow) {
+    String[] values = csvRow.split("\",\"");
+
+    values[0] = values[0].substring(1);
+
+    String last = values[values.length - 1];
+    values[values.length - 1] = last.substring(0, last.length() - 1);
+
+    return values;
+  }
+
+  private Object convertValue(String value, Schema.Field field) {
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.isNullable()) {
+      fieldSchema = fieldSchema.getNonNullable();
+    }
+
+    Schema.Type fieldSchemaType = fieldSchema.getType();
+
+    // empty string is considered null in csv, for all types but string.
+    if (value.isEmpty() && !fieldSchemaType.equals(Schema.Type.STRING)) {
+      return null;
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      switch (logicalType) {
+        case DATE:
+          // date will be in yyyy-mm-dd format
+          return Math.toIntExact(LocalDate.parse(value).toEpochDay());
+        case TIMESTAMP_MILLIS:
+          return Instant.parse(value).toEpochMilli();
+        default:
+          throw new UnexpectedFormatException(String.format("Field '%s' is of unsupported type '%s'",
+                                                            field.getName(), logicalType.getToken()));
+      }
+    }
+
+    switch (fieldSchemaType) {
+      case NULL:
+        return null;
+      case BOOLEAN:
+        return Boolean.parseBoolean(value);
+      case INT:
+        return Integer.parseInt(value);
+      case LONG:
+        return Long.parseLong(value);
+      case FLOAT:
+        return Float.parseFloat(value);
+      case DOUBLE:
+        return Double.parseDouble(value);
+      case BYTES:
+        return Byte.parseByte(value);
+      case STRING:
+        return value;
+    }
+
+    throw new UnexpectedFormatException(
```

Comment:
These are cdap types (not salesforce). The schema is generated automatically so there should no unknown types.

For actually converting salesforce to cdap types there is a different switch:
`  private static Schema fieldTypeToCdapSchemaType(FieldType fieldType) {
    switch(fieldType) {
      case _boolean:
        return Schema.of(Schema.Type.BOOLEAN);
      case _int:
        return Schema.of(Schema.Type.LONG);
      case _double:
      case currency:
      case percent:
        return Schema.of(Schema.Type.DOUBLE);
      case date:
        return Schema.of(Schema.LogicalType.DATE);
      case datetime:
        return Schema.of(Schema.LogicalType.TIMESTAMP_MILLIS);
      default:
        return Schema.of(Schema.Type.STRING);
    }
  }`

Salesforce has a lot of types like email, phone etc. which all just strings. So default is string. Also Salesforce allows creating custom types, which we also convert to string since don't have information about them.

Here is the types list of all Salesforce types:
`  string("string"),
  picklist("picklist"),
  multipicklist("multipicklist"),
  combobox("combobox"),
  reference("reference"),
  base64("base64"),
  _boolean("boolean"),
  currency("currency"),
  textarea("textarea"),
  _int("int"),
  _double("double"),
  percent("percent"),
  phone("phone"),
  id("id"),
  date("date"),
  datetime("datetime"),
  time("time"),
  url("url"),
  email("email"),
  encryptedstring("encryptedstring"),
  datacategorygroupreference("datacategorygroupreference"),
  location("location"),
  address("address"),
  anyType("anyType"),
  json("json"),
  complexvalue("complexvalue"),
  _long("long");`
>

Reply:
decided offline with Nitin that I will add a doc showing types mapping. All the types are supported.

Reply:
done